### ISSUE TITLE
HA fails on multiline SMTP status

### ISSUE THREAD

**Issue by alterratz** [haproxy/haproxy#1552]
### Detailed Description of the Problem

HA immediately drops the connection to the SMTP server when receiving a multiline reply as described in RFC 5321 Section 4.2.1

250-First line
250-Second line
250-234 Text beginning with numbers
250 The last line

After removing the first 3 lines SMTP alerts work, however it requires patching the mail server and violates the standard.

### Expected Behavior

Accept multiline Status messages

### Steps to Reproduce the Behavior

use this example python3 mail server to reproduce

```python
from datetime import datetime
import asyncore
from smtpd import SMTPServer

class EmlServer(SMTPServer):
    no = 0
    def process_message(self, peer, mailfrom, rcpttos, data, **kwargs):    
        filename = '%s-%d.eml' % (datetime.now().strftime('%Y%m%d%H%M%S'), self.no)
        print(filename)
        f = open(filename, 'wb')
        f.write(data)
        f.close
        print('%s saved.' % filename)
        self.no += 1

def run():
    s = EmlServer(('localhost', 10025), None)
    try:
        asyncore.loop()
    except KeyboardInterrupt:
        pass

if __name__ == '__main__':
    run()
```

### What is your configuration?

```haproxy
global
    description db2
    setenv PROXYNAME db2
    setenv PROXYPEER db1

    # process management and security
    chroot /home/***
    daemon
    master-worker
    user proxycheck
    group proxycheck
    #
    maxconn 512
    maxconnrate 32
    #
    ssl-default-bind-options   ssl-min-ver SSLv3
    ssl-default-server-options ssl-min-ver SSLv3
    #
    stats socket /var/run/haproxy/admin.sock mode 660 level admin expose-fd listeners

mailers mta
   mailer smtp1 127.0.0.1:10025

defaults
    log global
    mode http
    retries 2
    timeout client 30m
    timeout connect 4s
    timeout server 30m
    timeout check 5s
    # option log-health-checks
    # email settings
    email-alert mailers mta
    email-alert from alerts@"$PROXYNAME".mydomain.io
    email-alert to office@mydomain.io
    email-alert level info
    email-alert myhostname "$PROXYNAME".mydomain.io


# create an admin user and a view only user
userlist stats-auth
    group admin    users admin
    user  admin    insecure-password ***
    group readonly users status
    user  status  insecure-password ***

listen http_status:1080
    bind 127.0.0.1:1080
    stats enable
    stats uri /
    acl AUTH       http_auth(stats-auth)
    acl AUTH_ADMIN http_auth_group(stats-auth) admin
    stats http-request auth unless AUTH
    stats admin if AUTH_ADMIN

listen sslwrapper_database:25433
    bind *:25433 ssl crt /etc/haproxy/ssl/ssl.pem
    mode tcp
    server "$PROXYNAME" 127.0.0.1:5433 no-ssl maxconn 100

# database master
listen database:15432
    bind *:15432
    mode tcp
    tcp-request connection accept if { src 127.0.0.1 }
    tcp-request connection accept if { src 172.16.0.0/12 }
    tcp-request connection reject
    option httpchk /checkdb
    http-check expect status 200
    default-server inter 5s fall 3 rise 1 on-marked-down shutdown-sessions on-marked-up shutdown-backup-sessions
    server "$PROXYNAME" 127.0.0.1:5433 maxconn 100 check check-ssl verify none port 443
    server "$PROXYPEER" "$PROXYPEER.mydomain.io:25433" ssl maxconn 100 check check-ssl verify none port 443
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.7-2 2021/10/16 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.7.html
Running on: Linux 4.19.0-17-amd64 #1 SMP Debian 4.19.194-3 (2021-07-18) x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -ffile-prefix-map=/build/haproxy-SBafG1/haproxy-2.4.7=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=24).
Built with OpenSSL version : OpenSSL 1.1.1l  24 Aug 2021
Running on OpenSSL version : OpenSSL 1.1.1m  14 Dec 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

### Additional Information

If required I can proivide a tcp dump, but it is obvious that haproxy frops the connection when reveiving a multiline status reply from the MTA
(commenting out the offending lines in /usr/lib/python<x.y>/smtpd.py worked)

```
        self.extended_smtp = True
        #self.push('250-%s' % self.fqdn)
        #if self.data_size_limit:
        #    self.push('250-SIZE %s' % self.data_size_limit)
        #    self.command_size_limits['MAIL'] += 26
        #if not self._decode_data:
        #    self.push('250-8BITMIME')
        #if self.enable_SMTPUTF8:
        #    self.push('250-SMTPUTF8')
        #    self.command_size_limits['MAIL'] += 10
        self.push('250 HELP')
```

**Comment by alterratz** [comment]
Sorry for the formatting. I specifically used the <> formatting button

**Comment by TimWolla** [comment]
> Sorry for the formatting. I specifically used the <> formatting button

For multiline code blocks you need to use three backticks (```) for proper formatting. I've fixed that for you.

**Comment by lukastribus** [comment]
Multiline reply is ESMTP stuff, which we negotiate by sending `EHLO` instead of `HELO`.

Let's just use SMTP:

```
diff --git a/src/mailers.c b/src/mailers.c
index 3d01d7532..34eaa5bb6 100644
--- a/src/mailers.c
+++ b/src/mailers.c
@@ -195,7 +195,7 @@ static int enqueue_one_email_alert(struct proxy *p, struct server *s,
                goto error;

        {
-               const char * const strs[4] = { "EHLO ", p->email_alert.myhostname, "\r\n" };
+               const char * const strs[4] = { "HELO ", p->email_alert.myhostname, "\r\n" };
                if (!add_tcpcheck_send_strs(&alert->rules, strs))
                        goto error;
        }

```

**Comment by wtarreau** [comment]
Agreed, I wasn't even aware this was supported. @alterratz could you confirm that Lukas' fix works for you ?

**Comment by alterratz** [comment]
You learn something new every day  :-)   I was not aware of the significance of EHLO vs HELO.
However, sadly I don't have the resources at the moment to compile ant test HAproxy, but I verified with the python server implementation and when HELO is used instead of EHLO only a single status line is returned. I did this manually by commenting out the offending lines in the smtp server implementation and it worked nicely. So I'm 99.9% positive the provided patch will work.
Thank you very much for your help.
